hudi-flink-datasource/hudi-flink/src/test/java/org/apache/hudi/configuration/TestOptionsInference.java

```diff
@@ -69,6 +70,12 @@ void testSetupClientId() throws Exception {
     }
   }
 
+  @Test
+  void testAutoGenerateClient() {
+      Configuration conf = getConf();
+      OptionsInference.setupClientId(conf);
+      assertNotNull(conf.getString(FlinkOptions.WRITE_CLIENT_ID), "auto generate client failed!");
+  }
```

Comment:
Maybe we just set up the client id by default in `OptionsInference.setupClientId` ? I'm not sure what are we fixing here actually.

Reply:
This pr just fix the INIT_CLIENT_ID,in the ooc scene,if we don't set the config`write.client.id`,it will auto generate the CLIENT,and the INIT_CLIENT_ID doesn't null.

Reply:
> The `INIT_CLIENT_ID` is not null, it is an empty string, did you set up the `hoodie.write.concurrency.mode` as `OPTIMISTIC_CONCURRENCY_CONTROL` ?

 yes, i set up the `hoodie.write.concurrency.mode` as `OPTIMISTIC_CONCURRENCY_CONTROL`

Reply:
Can you explain again why we need this fix, when a Flink job was submitted, it is assigned a client id automically, this id is hold by this job with a hearbeat, when another job was submitted, it saw the heartbeat and would create a new client id, is this what you saw in you case?

Reply:
> Can you explain again why we need this fix, when a Flink job was submitted, it is assigned a client id automically, this id is hold by this job with a hearbeat, when another job was submitted, it saw the heartbeat and would create a new client id, is this what you saw in you case?

when a Flink job started,it is assigned a client id ,but the client id is a empty string,so the ckp_meta is shared by all writers.

Reply:
> so the ckp_meta is shared by all writers.

The writer should start a new client id if it saw the clienid with empty string is already occupied.

Reply:
> > so the ckp_meta is shared by all writers.
> 
> The writer should start a new client id if it saw the clienid with empty string is already occupied.

Actually,the writer doesn't generate a new client,because client id generated for the first time not empty.

Reply:
It should be a bug, the client still send heartbeat anyway for the INIT_CLIEN _ID: https://github.com/apache/hudi/blob/00d50e91abe24aba31daa2fe2806de5414f03c77/hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/util/ClientIds.java#L179

Reply:
> It should be a bug, the client still send heartbeat anyway for the INIT_CLIEN _ID:
> 
> https://github.com/apache/hudi/blob/00d50e91abe24aba31daa2fe2806de5414f03c77/hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/util/ClientIds.java#L179

Maybe,bug if the INIT_CLIENT_ID is empty,all writer will shared the ckp_meta,the risk of concurrent modification will occupy.

Reply:
> all writer will shared the ckp_meta

How could that happen then providing the last client already sent the heartbeat?

Reply:
Yeah, when a new client id is generated, the client would immediately sent a heart beat to the table, while other clients see the heartbeat, they skip the occupied hearbeated ids, can you help to check whether that works as expected?

Reply:
> Yeah, when a new client id is generated, the client would immediately sent a heart beat to the table, while other clients see the heartbeat, they skip the occupied hearbeated ids, can you help to check whether that works as expected?

OK